pkg/httpclient/client_test.go

```diff
@@ -0,0 +1,247 @@
+		// the total time of the request should be 300ms
+		// the time between retries should be 100ms
+		{
+			name: "linear retry when no codes specified but the request times out",
+			srv:  newTestSrv(t, http.StatusOK, []byte("OK"), 450),
+
+			ctxDeadline:     5 * time.Second,
+			backoff:         httpclient.BackoffLinear,
+			maxRetries:      3,
+			perRetryTimeout: 400 * time.Millisecond,
+			initialBackoff:  100 * time.Millisecond,
+
+			wantRetry:         true,
+			wantClientTimeout: true,
+			wantErr:           errors.New("i/o timeout"),
+		},
+
+		// the request should be retried linearly every 500ms a max of 3 times.
+		// the total time of the request should be 1000ms
+		// the time between retries should be 500ms
+		{
+			name: "linear retry",
+			srv:  newTestSrv(t, http.StatusInternalServerError, nil, 0),
+
+			codes:           []int{http.StatusInternalServerError},
+			ctxDeadline:     5 * time.Second,
+			backoff:         httpclient.BackoffLinear,
+			maxRetries:      3,
+			perRetryTimeout: 5 * time.Second,
+			initialBackoff:  500 * time.Millisecond,
+
+			wantRetry:       true,
+			wantCode:        http.StatusInternalServerError,
+			wantTotalWait:   1000 * time.Millisecond,
+			wantWaitBetween: 500 * time.Millisecond,
+			wantCalled:      3,
+			wantBody:        []byte{},
+		},
+		// the request should be retried exponentialy starting from 100ms i.e 100ms * (2 ^ attempt) for a max of 3 times
+		// the total time of the request should be:
+		// 		total = 0ms  // attempt 1
+		//      total += 100ms * (2 ^ 2) = 400ms  // attempt 2
+		// 		total += 100ms * (2 ^ 3) = 800ms // attempt 3
+		//      total = 1200
+		// the time between retries ( we will only check the last retry attempt) = 800ms
+		{
+			name: "exponential retry",
+			srv:  newTestSrv(t, http.StatusInternalServerError, nil, 0),
+
+			codes:           []int{http.StatusInternalServerError},
+			ctxDeadline:     5 * time.Second,
+			backoff:         httpclient.BackoffExponential,
+			maxRetries:      3,
+			perRetryTimeout: 5 * time.Second,
+			initialBackoff:  100 * time.Millisecond,
+
+			wantRetry:       true,
+			wantWaitBetween: 800 * time.Millisecond,
+			wantCalled:      3,
+			wantCode:        http.StatusInternalServerError,
+			wantTotalWait:   1200 * time.Millisecond,
+			wantBody:        []byte{},
+		},
+	}
+
+	for i := range tb {
+		tt := tb[i]
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			ctx, cancel := context.WithTimeout(context.Background(), tt.ctxDeadline)
+			defer cancel()
+
+			srv := httptest.NewServer(http.HandlerFunc(tt.srv.Handle))
+			defer srv.Close()
+
+			retryPolicy := httpclient.NewRetryPolicy(tt.perRetryTimeout, tt.maxRetries).
+				WithCodes(tt.codes...).WithBackoff(tt.backoff, tt.initialBackoff)
+
+			tmp := srv.Client()
+			tmp.Timeout = tt.perRetryTimeout
+			client := httpclient.NewHTTPClient(httpclient.Config{
+				CustomClient: tmp,
+				RetryPolicy:  retryPolicy,
```

Comment:
I think it's good to have a general level configuration, but I would also like to have one at the endpoint level, since it might happen that we inject the httpclient into a package and it's used for different endpoints with different retry configurations